### ISSUE TITLE
target-size: false positive when translucent target is inside a positioned "fake" stacking context

### ISSUE THREAD

**Issue by dbjorge** [dequelabs/axe-core#4350]
### Product

axe-core

### Product Version

4.8.4

### Latest Version

- [X] I have tested the issue with the latest version of the product

### Issue Description

We got a customer report today of a target-size false positive (see minimal repro below). This turned out to be because the stacking context tracking used by the target-offset check has an issue where it only has a partial understanding of the rules for elements with `opacity` < 1. Per https://www.w3.org/TR/css-color-3/#transparency:

>Since an element with opacity less than 1 is composited from a single offscreen image, content outside of it cannot be layered in z-order between pieces of content inside of it. For the same reason, implementations must create a new stacking context for any element with opacity less than 1. If an element with opacity less than 1 is not positioned, then it is painted on the same layer, within its parent stacking context, as positioned elements with stack level 0. If an element with opacity less than 1 is positioned, the ‘z-index’ property applies as described in [[CSS21]](https://www.w3.org/TR/css-color-3/#ref-CSS21), except that if the used value is ‘auto’ then the element behaves exactly as if it were ‘0’. See [section 9.9](https://www.w3.org/TR/CSS21/visuren.html#layers) and [Appendix E](https://www.w3.org/TR/CSS21/zindex.html) of [[CSS21]](https://www.w3.org/TR/css-color-3/#ref-CSS21) for more information on stacking contexts. The rules in this paragraph do not apply to SVG elements, since SVG has its own [rendering model](https://www.w3.org/TR/SVG11/render.html) ([[SVG11]](https://www.w3.org/TR/css-color-3/#ref-SVG11), Chapter 3).

Currently, `create-grid.js`'s `isStackingContext` does understand that opacity < 1 should results in a stacking context being created, but `getStackLevel` doesn't understand the rule buried in the middle about non-positioned translucent elements being painted alongside *positioned* elements. The code in `createStackingOrder` that distinguished fake and true stacks also doesn't understand to treat opacity-induced stacking contexts as "true" contexts, which the spec suggests we're supposed to do even when we draw them in POSITION_LEVEL order.

In the repro case below (and the internally-reported customer site), this results in `hasVisualOverlap` (via `visuallySort`) incorrectly determining that the positioned parent completely overlaps and obscures the translucent child, which isn't what happens in browsers in practice. This in turn results in `getTargetRects` returning an empty list such that #4194 gets triggered (which in turn results in the `target-size` false positive).

I titled this bug according to the user-facing impact, but let's have this one just track the bad stack context behavior; #4194 already tracks the bad `getTargetRects`/`targetOffsetEvaluate` behavior.

#### Expectation
Repro case below shouldn't violate the `target-size` rule.

#### Actual
It does.

#### How to Reproduce
```html
<div style="position: absolute; transform: scale(1)">
  <button id="t1">first tab</button>
  <button id="t2" style="opacity: .6">second tab</button>
</div>
```

#### Additional context
- https://www.w3.org/TR/css-color-3/#transparency
- #4194
- #4295

**Comment by dbjorge** [comment]
Instructions for testing whether a possible repro is actually hitting this bug or not:

1. Open the site in question in the reproing state
2. Right click the element in question and select "Inspect"
3. With the element selected in the dev tools panel, paste the following snippet into the dev tools console:
```js
function testForIssue4350(elementOrSelector) {
  const elm = (typeof elementOrSelector === 'string') ? document.querySelector(elementOrSelector) : elementOrSelector;
  const elmRect = elm.getBoundingClientRect();
  const elmX = elmRect.left + elmRect.width / 2;
  const elmY = elmRect.top + elmRect.height / 2;

  const prettyStack = elms => JSON.stringify(elms.map(elm => axe.utils.getSelector(elm)), null, 2);

  axe.teardown()
  try {
    axe.setup()
    const axeStack = prettyStack(axe.commons.dom.getElementStack(elm));
    const rawUaStack = document.elementsFromPoint(elmX, elmY);
    const uaStack = prettyStack(rawUaStack);

    if (rawUaStack[0] !== elm) {
      console.log('The user agent thinks this element is obscured.');
    }
    if (axeStack !== uaStack) {
      console.log('Axe and UA stacks do not match. This is probably a repro of axe-core#4350.');
      console.log(`Axe stack:\n${axeStack}\nUA stack:\n${uaStack}`);
    } else {
      console.log('Axe and UA stacks match. This is probably *not* a repro of axe-core#4350.');
      console.log(`Stack:\n${axeStack}`);
    }
  } finally {
    axe.teardown()
  }
}
testForIssue4350($0)
```
4. Observe the output on the console

**Comment by chutchins25** [comment]
Fixed on `develop` by #5214.

This issue tracks the stack-context bug specifically (the FP itself was #4194, now closed). Testing the repro with the diagnostic from the comment above — comparing `getElementStack` against `document.elementsFromPoint`:

- **Before #5214:** axe stack `[div, #t2, html]` vs UA stack `[#t2, div, html]` — mismatch. axe ranks the positioned parent above the translucent child, exactly as described.
- **On develop (after #5214):** axe stack `[#t2, div, html]` == UA stack — match.

#5214 ("fix various stacking context bugs") reworked `createStackingOrder` / `getStackLevel` in `create-grid.js` so opacity-induced stacking contexts are treated as real stacks and sorted at the positioned layer — the spec rule quoted above. It's the same fix that closed the sibling #4629. `target-size` no longer flags `#t2`.

The draft PR #4351 (Closes #4350) looks superseded by this — probably safe to close it too. Safe to close.
